### PR TITLE
bathymetry plugin ready to review

### PR DESCRIPTION
**Pull request by woensug-choi**
Error occurred with the original bathy_dave_plugin branch due to large bathymetry source file inside committed stacks.

Large bathymetry source files are removed from plugin source codes. (Still, 40 MB-ish meshlab portable is included)

This new branch is ready to review. Please check following the tutorial wiki page. (Be warned, following tutorial may take almost two hours)

[Bathymetry plugin tutorial link](https://github.com/Field-Robotics-Lab/dave/wiki/Bathymetry-Plugin)

### CODE REVIEW

**Review by M1chaelM — COMMENTED** [review]
The dependency on tools that require an additional 1.5 hours to download and compile is something we need to work through before merging this code, since it greatly increases the barrier to entry for building and using our code. In addition, compiling Python packages from source is generally not ideal since it is easy to end up with version conflicts in your environment. If there's really no package manager that supports gdal and pdal, we need at least to give instructions for installing them into a virtual environment, docker image, or something that will allow the user to find, update, and / or remove them later.

As a first step, I'd like to work with @caguero and @mabelzhang to make sure there aren't alternatives that are already. We discussed this briefly at last week's meeting but we haven't had time to sync up yet.

**Review by caguero — COMMENTED** [review]
Is it accurate to say that the process described in the tutorial converts a DEM file into one or multiple `.obj` files that can be loaded into Gazebo?

If so, this is an interesting way to do it because Gazebo has a relatively good support for meshes, so this should work besides the long installation time required.

Alternatively, Gazebo supports loading directly a DEM file. See [this tutorial](http://gazebosim.org/tutorials?tut=dem&cat=build_world). The tutorial is a bit old, some tools may not be available but the DEM support should work. It's been used by NASA in other project as well. In this approach, you don't do any conversion, you load directly your DEM file.

**Comment by woensug-choi** [comment]
@caguero 
## Conversion process
tif bathymetry (downloaded from NOAA) -> convert to ply format (gdal/pdal) -> convert to obj format (meshlab portable) -> generate mesh sdfs with objs (mkbathy.sh)

**Comment by caguero** [comment]
> @caguero
> Conversion process :
> `.tif` bathymetry (downloaded from NOAA, this example case 200MB) -> convert to `.ply` format (gdal/pdal) -> convert to `.obj` format (meshlab portable) -> generate separated mesh model `.sdf` with objs (mkbathy.sh)

Thanks! As I said before, if the 1.5 hour prerequisite is a showstopper, I'd suggest to test if you can load directly the DEM into Gazebo without doing any conversion. Otherwise it looks like a valid way to load the elevation.

**Comment by woensug-choi** [comment]
@caguero 
I just understood that tiff is DEM! Not sure why the original code had a complex conversion process. Will ask Michael Jakuba before proceeding forward.  Thank you!

**Comment by caguero** [comment]
> @caguero
> I just understood that tiff is DEM! Not sure why the original code had a complex conversion process. Will ask Michael Jakuba before proceeding forward. Thank you!

Correct, DEM is the generic name and `tiff` is a specific format to capture a DEM but not the only format. `gdal` deals with most of the known formats.

**Comment by caguero** [comment]
The notes in this project might be useful as well.

https://github.com/osrf/uav_testing

**Comment by woensug-choi** [comment]
According to the last week's discussions, the PR is updated to exclude data conversion to a separate standalone repository.
In this PR, it only includes bathymetry tiles that are already been converted with overlapping regions and namings just to reproduce the video clip on the [tutorial wiki](https://github.com/Field-Robotics-Lab/dave/wiki/Bathymetry-Plugin).

**This is ready to merge.** 👍

**Comment by M1chaelM** [comment]
> According to the last week's discussions, the PR is updated to exclude data conversion to a separate standalone repository.
> In this PR, it only includes bathymetry tiles that are already been converted with overlapping regions and namings just to reproduce the video clip on the [tutorial wiki](https://github.com/Field-Robotics-Lab/dave/wiki/Bathymetry-Plugin).
> 
> **This is ready to merge.** 👍

Thank you! I will try to review this today.

**Comment by woensug-choi** [comment]
Getting error msg launching with `roslaunch nps_uw_sensors_gazebo depth_camera_sonar_basic.launch`
```bash
process[depth_and_normals_to_sonar-5]: started with pid [4391]
/usr/bin/env: ‘python\r’: No such file or directory
process[rqt_image_view_image_beams-6]: started with pid [4392]
[depth_and_normals_to_sonar-5] process has died [pid 4391, exit code 127, cmd /home/eric/uuv_ws/src/nps_uw_sensors_gazebo/scripts/sonar_ros.py __name:=depth_and_normals_to_sonar __log:=/home/eric/.ros/log/8a5c2930-fba5-11ea-9dd5-bcee7b8ddb04/depth_and_normals_to_sonar-5.log].
log file: /home/eric/.ros/log/8a5c2930-fba5-11ea-9dd5-bcee7b8ddb04/depth_and_normals_to_sonar-5*.log
```
according to [this thread](https://askubuntu.com/questions/896860/usr-bin-env-python3-r-no-such-file-or-directory)
`The problem are your line ending characters. Your file was created or edited on a Windows system and uses Windows/DOS-style line endings (CR+LF), whereas Linux systems like Ubuntu require Unix-style line endings (LF).`
using package dos2unix with command `dos2unix *.py` fixed the problem which is installed with `sudo apt install dos2unix`

**Comment by M1chaelM** [comment]
@woensug-choi  After I launch and unpause the simulation, the environment remains gray and blank. I get a message saying:
```
[Dbg] [bathy_plugin.cc:140] Loading bathymetry plugin for rexrov...
[Dbg] [bathy_plugin.cc:209] Update bathy!
[Dbg] [bathy_plugin.cc:317] lonc/col: -70.675/2 latc/row: 41.519/1
[Dbg] [bathy_plugin.cc:331] Inserted model: ncei19.R_-70.675_-70.663_41.519_41.529
[Wrn] [ModelDatabase.cc:340] Getting models from[http://models.gazebosim.org/]. This may take a few seconds.
[Msg] Waiting for model database update to complete...
```
Then it pauses for a while. Next, get some warnings and errors:
```
[Wrn] [FuelModelDatabase.cc:248] [Wrn] [FuelModelDatabase.cc:248] URI not supported by Fuel [model://ncei19.R_-70.675_-70.663_41.519_41.529.epsg26987/meshes/ncei19.R_-70.675_-70.663_41.519_41.529.epsg26987.objURI not supported by Fuel [URI not supported by Fuel [URI not supported by Fuel [model://ncei19.R_-70.675_-70.663_41.519_41.529.epsg26987/meshes/ncei19.R_-70.675_-70.663_41.519_41.529.epsg26987.obj]
[Wrn] [SystemPaths.cc:464] File or path does not exist [[Wrn] [SystemPaths.cc:464] File or path does not exist [[Wrn] [SystemPaths.cc:464] ""[1;"] [33mmodel://ncei19.R_-70.675_-70.663_41.519_41.529.epsg26987/meshes/ncei19.R_-70.675_-70.663_41.519_41.529.epsg26987.obj"][1;] [33mmodel://ncei19.R_-70.675_-70.663_41.519_41.529.epsg26987/meshes/ncei19.R_-70.675_-70.663_41.519_41.529.epsg26987.obj[1;33m
[1;]33m
[Err] [Visual.cc:3043] [1;No mesh specified
[Err] [MeshShape.cc:60] 31mNo mesh specified
[Wrn] [FuelModelDatabase.cc:248] URI not supported by Fuel [model://ncei19.R_-70.675_-70.663_41.519_41.529.epsg26987/meshes/ncei19.R_-70.675_-70.663_41.519_41.529.epsg26987.obj]
[Wrn] [SystemPaths.cc:464] File or path does not exist [""] [model://ncei19.R_-70.675_-70.663_41.519_41.529.epsg26987/meshes/ncei19.R_-70.675_-70.663_41.519_41.529.epsg26987.obj]
[Err] [Visual.cc:3043] No mesh specified
[Wrn] [FuelModelDatabase.cc:248] URI not supported by Fuel [model://ncei19.R_-70.675_-70.663_41.519_41.529.epsg26987/meshes/ncei19.R_-70.675_-70.663_41.519_41.529.epsg26987.obj]
[Wrn] [SystemPaths.cc:464] File or path does not exist [""] [model://ncei19.R_-70.675_-70.663_41.519_41.529.epsg26987/meshes/ncei19.R_-70.675_-70.663_41.519_41.529.epsg26987.obj]
[Wrn] [FuelModelDatabase.cc:248] URI not supported by Fuel [model://ncei19.R_-70.675_-70.663_41.519_41.529.epsg26987/meshes/ncei19.R_-70.675_-70.663_41.519_41.529.epsg26987.obj]
[Wrn] [SystemPaths.cc:464] File or path does not exist [""] [model://ncei19.R_-70.675_-70.663_41.519_41.529.epsg26987/meshes/ncei19.R_-70.675_-70.663_41.519_41.529.epsg26987.obj]
[Err] [Visual.cc:3043] No mesh specified
[Wrn] [FuelModelDatabase.cc:248] URI not supported by Fuel [model://ncei19.R_-70.675_-70.663_41.519_41.529.epsg26987/meshes/ncei19.R_-70.675_-70.663_41.519_41.529.epsg26987.obj]
[Wrn] [SystemPaths.cc:464] File or path does not exist [""] [model://ncei19.R_-70.675_-70.663_41.519_41.529.epsg26987/meshes/ncei19.R_-70.675_-70.663_41.519_41.529.epsg26987.obj]
[Dbg] [bathy_plugin.cc:209] Update bathy!
```
I'm running in a Docker container; not sure whether that might have an effect. Let me know if there's anything I can do to provide more information.

**Comment by woensug-choi** [comment]
Thank you for the review. It took me forever to find out what the problem is.
The dave project's `.gitignore` had `obj` files to be excluded in the commit. Changed .gitignore and added demo obj files.

**Comment by M1chaelM** [comment]
Great, thanks. This is running now. I'll review it.

File: bathy_dave_plugin/launch/robots.launch
Comment (M1chaelM) [inline bathy_dave_plugin/launch/robots.launch:9]: I made some minor changes to the initial position so we can move the ROV more quickly to a position that triggers the new tile.

File: bathy_dave_plugin/world/bathy_dave.world
Comment (M1chaelM) [inline bathy_dave_plugin/world/bathy_dave.world:11]: I also moved the camera so the ROV is visible when the simulation launches. I'm not sure whether this will work for you: I'm finding that the "clip" settings have no effect and I'm unable to set the initial zoom of the camera. This could be because my version of Gazebo is not the very latest, so if you find that this position doesn't show the ROV when you run it, please change it back.

**Review by M1chaelM — APPROVED** [review]
All looks good--I only left two very minor comments about the initial launch conditions. 
* I will make some edits to the tutorial to mention that the initial screen is blank gray and they should be prepared to wait for 15 seconds or so for it to load. 
* I also have a question about the tiles:
I see some places where the texture seems to have blank spaces in it:
crack_in_spacetime
I checked to make sure, and it seems that these are "real" holes. That is, they have physical as well as visual properties and it is possible to drive the vehicle through them:
into_the_wormhole

Do you know what causes this? It seems like we should add some explanation to the tutorial.

Since none of the above represent major issues I'm going to approve the merge request. However, if you think we need to create an issue for the "cracks" in the tiles, please make one for milestone 3.0.

Thank you!

File: bathy_dave_plugin/world/bathy_dave.world
Comment (woensug-choi) [reply bathy_dave_plugin/world/bathy_dave.world:11]: Works perfect.

**Comment by woensug-choi** [comment]
"cracks" on tiles are originated from NOAA tiff files. Can't find where exactly I saw the same issues but seen that somewhere else. In order to smooth and sweep those holes out, delicate work should be performed on Meshlab to hand-stitch them. Only if needed.

If we were to generate some sort of interpolation sweeping methods, it should be on separated 'Bathymetry Converter' repository. I don't think it should be covered in the dave project. 

Leaving this PR left for you to click the merge button so that you can read this last comment.

**Comment by M1chaelM** [comment]
> Leaving this PR left for you to click the merge button so that you can read this last comment.

@woensug-choi Got it. Thanks. Merging.

**Comment by crvogt** [comment]
Not yet deleting branch as it is relevant to issue Field-Robotics-Lab/Bathymetry_Converter#1

**Pull request merged**
